core/services/offchainreporting/peerstore.go

```diff
@@ -2,21 +2,153 @@ package offchainreporting
 
 import (
 	"context"
-	"database/sql"
+	"time"
 
-	sqlds "github.com/ipfs/go-ds-sql"
-	pgqueries "github.com/ipfs/go-ds-sql/postgres"
+	"github.com/jinzhu/gorm"
+	p2ppeer "github.com/libp2p/go-libp2p-core/peer"
 	p2ppeerstore "github.com/libp2p/go-libp2p-core/peerstore"
-	p2ppeerstoreds "github.com/libp2p/go-libp2p-peerstore/pstoreds"
+	"github.com/libp2p/go-libp2p-peerstore/pstoremem"
+	ma "github.com/multiformats/go-multiaddr"
+	"github.com/pkg/errors"
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/services/postgres"
+	"github.com/smartcontractkit/chainlink/core/utils"
 )
 
-const tableName = "p2p_peerstore"
+type (
+	p2pPeer struct {
+		ID        p2ppeer.ID
+		Addr      ma.Multiaddr
+		JobID     int32
+		CreatedAt time.Time
+		UpdatedAt time.Time
+	}
 
-// NewPeerstore creates a new database-backed peerstore
-// NOTE: You can get sql.DB from store with store.DB.DB()
-func NewPeerstore(ctx context.Context, db *sql.DB) (p2ppeerstore.Peerstore, error) {
-	queries := pgqueries.NewQueries(tableName)
-	datastore := sqlds.NewDatastore(db, queries)
-	opts := p2ppeerstoreds.DefaultOpts()
-	return p2ppeerstoreds.NewPeerstore(ctx, datastore, opts)
+	Pstorewrapper struct {
+		utils.StartStopOnce
+		Peerstore     p2ppeerstore.Peerstore
+		jobID         int32
+		db            *gorm.DB
+		writeInterval time.Duration
+		ctx           context.Context
+		ctxCancel     context.CancelFunc
+		chDone        chan struct{}
+	}
+)
+
+// NewPeerstoreWrapper creates a new database-backed peerstore wrapper scoped to the given jobID
+// Multiple peerstore wrappers should not be instantiated with the same jobID
+func NewPeerstoreWrapper(ctx context.Context, db *gorm.DB, writeInterval time.Duration, jobID int32) (*Pstorewrapper, error) {
+	ctx, cancel := context.WithCancel(ctx)
+
+	return &Pstorewrapper{
+		utils.StartStopOnce{},
+		pstoremem.NewPeerstore(),
+		jobID,
+		db,
+		writeInterval,
+		ctx,
+		cancel,
+		make(chan struct{}),
+	}, nil
+}
+
+func (p *Pstorewrapper) Start() error {
+	if !p.OkayToStart() {
+		return errors.New("cannot start")
+	}
+	err := p.readFromDB()
+	if err != nil {
+		return errors.Wrap(err, "could not start peerstore wrapper")
+	}
+	go p.dbLoop()
+	return nil
+}
+
+func (p *Pstorewrapper) dbLoop() {
+	defer close(p.chDone)
+	ticker := time.NewTicker(p.writeInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-p.ctx.Done():
```

Comment:
looks like we won't call p.Peerstore.Close() if a user cancels the parent context (would need to explicitly call Close). Would we want to also call p.Peerstore.Close() upon parent context cancel?

Reply:
Good atch